daemon/algod/api/server/v2/simulate.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package v2
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/daemon/algod/api/server/v2/generated"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/verify"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/pkg/errors"
+)
+
+// ==============================
+// > Simulation Ledger
+// ==============================
+
+// LedgerForSimulator is a ledger interface for the simulator.
+type LedgerForSimulator interface {
+	ledger.DebuggerLedgerForEval
+}
+
+type apiSimulatorLedgerConnector struct {
+	LedgerForAPI
+	hdr bookkeeping.BlockHeader
+}
+
+// Latest is part of the LedgerForSimulator interface.
+// We override this to use the set hdr to prevent racing with the network
+func (l apiSimulatorLedgerConnector) Latest() basics.Round {
+	return l.hdr.Round
+}
+
+// BlockHdr is part of the LedgerForSimulator interface.
+// We override this to use the set hdr to prevent racing with the network
+func (l apiSimulatorLedgerConnector) BlockHdr(round basics.Round) (bookkeeping.BlockHeader, error) {
+	if round != l.Latest() {
+		err := fmt.Errorf(
+			"BlockHdr() evaluator called this function for the wrong round %d, "+
+				"latest round is %d",
+			round, l.Latest())
+		return bookkeeping.BlockHeader{}, err
+	}
+
+	return l.LedgerForAPI.BlockHdr(round)
+}
+
+// BlockHdrCached is part of the LedgerForSimulator interface.
+func (l apiSimulatorLedgerConnector) BlockHdrCached(round basics.Round) (bookkeeping.BlockHeader, error) {
+	return l.BlockHdr(round)
+}
```

Comment:
This won't cut it for the `block` opcode. It can look backward beyond `Latest()`